controllers/cluster_controller.go

```diff
@@ -287,3 +296,80 @@ func splitMachineList(list *clusterv1.MachineList) (*clusterv1.MachineList, *clu
 	}
 	return controlplanes, nodes
 }
+
+func (r *ClusterReconciler) reconcileControlPlaneInitialized(ctx context.Context, cluster *clusterv1.Cluster) error {
+	if cluster.Status.ControlPlaneInitialized {
+		return nil
+	}
+
+	machines, err := r.getMachinesInCluster(ctx, cluster.Namespace, cluster.Name)
+	if err != nil {
+		r.Log.Error(err, "error getting machines in cluster", "cluster", cluster.Name)
+		return err
+	}
+
+	for _, m := range machines {
+		if util.IsControlPlaneMachine(m) && m.Status.NodeRef != nil {
+			cluster.Status.ControlPlaneInitialized = true
+			return nil
+		}
+	}
+
+	return nil
+}
+
+// getMachinesInCluster returns all of the Machine objects that belong to the cluster
+func (r *ClusterReconciler) getMachinesInCluster(ctx context.Context, namespace, name string) ([]*clusterv1.Machine, error) {
```

Comment:
@vincepri @ncdc  I created this function similar to this in machine controller. should move this as a common function elsewhere?
https://github.com/kubernetes-sigs/cluster-api/blob/2a5557304210e1e21c2bc33c1fffda4187e276f8/controllers/machine_controller.go#L264-L284

Reply:
Yes please! You can change the signature so it takes in the `Client`. As for where to put it... I'm trying to avoid generic files/packages like `util` or `helpers`. Even though I just said let's avoid "helpers", maybe we put it in `controllers/machine_helpers.go`? @vincepri wdyt?

Reply:
Either way should be fine, I'm also ok adding it in `util/util.go` for now and remove it later, there are a bunch of other similar functions that take a Controller Runtime Client and return something, so this wouldn't be anything new.

If we want to make it private, `controllers/helpers.go` or `controllers/machine_helpers.go` as you suggested sounds good.